sdk/src/distributed/worker/pool.rs

```diff
@@ -0,0 +1,345 @@
+
+                    results.push((request_idx, response));
+
+                    // If there is another task to redistribute, do it right away
+                    if !requests_to_redistribute.is_empty() {
+                        let request_idx = requests_to_redistribute.pop_front().unwrap();
+                        let request = requests[request_idx].clone();
+                        let worker = Arc::clone(self.workers.get(&worker_idx).unwrap());
+
+                        log::info!(
+                            "Sp1 Distributed: Redistributing task to worker {}",
+                            worker_idx
+                        );
+
+                        set.spawn(async move {
+                            (
+                                request_idx,
+                                worker_idx,
+                                worker.write().await.request(request).await,
+                            )
+                        });
+                    } else {
+                        available_workers.push_back(worker_idx);
+                    }
+                }
+                Err(_e) => {
+                    log::warn!("Sp1 Distributed: Worker {} failed", request_idx);
+
+                    failed_workers.push(worker_idx);
+
+                    if failed_workers.len() == self.workers.len() {
+                        return Err(WorkerError::AllWorkersFailed);
+                    }
+
+                    // If no other workers finished, push back the request to be picked up when a
+                    // worker is available
+                    if available_workers.is_empty() {
+                        requests_to_redistribute.push_back(request_idx);
+
+                        continue;
+                    }
+
+                    let worker_id = available_workers.pop_front().unwrap();
+                    let request = requests[request_idx].clone();
+                    let worker = Arc::clone(self.workers.get(&worker_id).unwrap());
+
+                    log::info!(
+                        "Sp1 Distributed: Redistributing task to worker {}",
+                        worker_id
+                    );
+
+                    set.spawn(async move {
+                        (
+                            request_idx,
+                            worker_id,
+                            worker.write().await.request(request).await,
+                        )
+                    });
+                }
+            }
+        }
+
+        results.sort_by_key(|(i, _)| *i);
+
+        let results = results.into_iter().map(|(_, response)| response).collect();
+
+        // removing the failed workers from the pool
+        for worker_id in failed_workers {
+            self.workers.remove(&worker_id);
+        }
+
+        Ok(results)
+    }
+
+    async fn spawn_workers() -> Result<BTreeMap<usize, Arc<RwLock<WorkerSocket>>>, WorkerError> {
+        let ip_list_string = std::fs::read_to_string("distributed.json")
```

Comment:
Should this be a config file with a default to distributed.json?

We probably want to offer people to have:
`distributed-mainnet.json`
`distributed-devnet.json`
`distributed-myboosterrollup1.json`
to manage multiple infra from the same repo

This can be a future refactor.

Reply:
This is indeed a good idea. We should also offer to change to path of such files via some config